### PR TITLE
feat(metrics): add metrics for ban and been_baned

### PR DESCRIPTION
**Pull request by chanhsu001**
### What problem does this PR solve?

Add metrics for network ban peer or  been banned, help to measure network partition.

### What is changed and how it works?

What's Changed:
ban peer operation ocurred at sync and tx_pool module, but callee function is at network module, network.rs
we add ban_peer and reason info. in metrics.

- PR to update `owner/repo`:

### Check List 

Tests 

- No code ci-runs-only: [ quick_checks,linters ]
- test manually

### Release note 

```release-note
None: Exclude this PR from the release note.
```

### CODE REVIEW

**Comment by chanhsu001** [comment]
I just want to write a test but don't know how to setup promtheus backend...

File: network/src/protocols/disconnect_message.rs [hunk 1]
```diff
@@ -27,6 +28,36 @@ impl ServiceProtocol for DisconnectMessageProtocol {
                 "Received disconnect message from peer={}: {}",
                 session_id, message
             );
+
+            // check if baned message and metric banned info
+            if message.starts_with("Ban for") && message.ends_with("seconds") {
```
Comment (doitian) [inline network/src/protocols/disconnect_message.rs:33]: Refer to the `NetworkState::ban_session` here to help code readers double check the message.

Comment (doitian) [reply network/src/protocols/disconnect_message.rs:33]: We can add ban reason in the message in `NetworkState::ban_session`. Here we should handle backward compatibility. It's OK when the ban message has no reason, but if it does, we can record it.

**Review by doitian — CHANGES_REQUESTED** [review]

File: network/src/protocols/disconnect_message.rs [hunk 2]
```diff
@@ -63,3 +79,33 @@ impl ServiceProtocol for DisconnectMessageProtocol {
...
+    // parse disconnect message and return peer_ip, ban seconds and reason(if exists)
+    // refer to NetworkState::ban_session() for message format
+    fn parse_disconnect_message(
+        &self,
+        session_id: SessionId,
+        message: &str,
+    ) -> (String, String, String) {
+        let addr = self.0.with_peer_registry(|reg| {
+            reg.get_peer(session_id)
+                .filter(|peer| !peer.is_whitelist)
```
Comment (quake) [inline network/src/protocols/disconnect_message.rs:93]: why filter out whitelist?

File: network/src/protocols/disconnect_message.rs [hunk 3]
```diff
@@ -63,3 +79,33 @@ impl ServiceProtocol for DisconnectMessageProtocol {
...
+                .map(|peer| peer.connected_addr.clone())
+                .map_or(String::default(), |maddr| maddr.to_string())
```
Comment (quake) [inline network/src/protocols/disconnect_message.rs:95]:
```suggestion
                .map(|peer| peer.connected_addr.to_string())
                .unwrap_or_default()
```

File: network/src/protocols/disconnect_message.rs [hunk 4]
```diff
@@ -63,3 +79,33 @@ impl ServiceProtocol for DisconnectMessageProtocol {
...
+        });
+
+        let mut ban_seconds = String::new();
+        if let Some(start) = message.find(char::is_numeric) {
```
Comment (quake) [inline network/src/protocols/disconnect_message.rs:99]: I don't think this code block is working for a ban message with any numeric char in the reason section, for example, with status code:
`Ban for 15 seconds, reason: HeadersIsInvalid(415)`

**Review by quake — DISMISSED** [review]

File: network/src/protocols/disconnect_message.rs (hunk 4 above)
Comment (chanhsu001) [reply network/src/protocols/disconnect_message.rs:99]: alright, i will correct it.

File: network/src/protocols/disconnect_message.rs (hunk 2 above)
Comment (chanhsu001) [reply network/src/protocols/disconnect_message.rs]: filter of whitelist is unnecessary and removed...

**Comment by chanhsu001** [comment]
After I write testcase to confirm metrics export to Prometheus, i will update result here.

**Review by doitian — APPROVED** [review]

File: network/src/protocols/disconnect_message.rs (hunk 3 above)
Comment (quake) [reply network/src/protocols/disconnect_message.rs]: this review comment hasn't been resolved, please check again.

Comment (chanhsu001) [reply network/src/protocols/disconnect_message.rs]: i miss that part. thanks for your patience.

**Comment by chanhsu001** [comment]
Close it now, ban/banned operation to inspect network partition may solve by other method.

**Pull request merged**